### DISCUSSION TITLE
Missing AdditionalFiles when building a project using a Azure DevOps pipeline

### DISCUSSION THREAD

**Discussion by Benny56** [dotnet/roslyn#51903] [Q&A] [answered]
I have a Xamarin Forms project which uses an xml file to control the source generator as described in the cookbook. The source generator is published as a NuGet package. Locally all builds succeed. 

When it comes to use Azure DevOps CI/CD the source generator is invoked during the build process, but it does not find the xml file, means context.AdditionalFiles.Length == 0 instead of 1 in my case.

My csproj is very simple:

  <ItemGroup>
    <AdditionalFiles Include="MobileAppGeneratorSettings.xml" />
    <PackageReference Include="MobileAppGenerator" Version="2.0.0" />
  </ItemGroup>

What is missing?

I have attached the build log. Please find "Error MOBILEAPPGEN002" in line 294. The source generator outputs this diagnostic message when context.AdditionalFiles.Length == 0.

[BuildLog.txt](https://github.com/dotnet/roslyn/files/6147511/BuildLog.txt)

**Comment by jmarolf** [comment]
@Benny56 can you record and upload a binary log of your build? That should give enough information that we could pin point what is going wrong.

**Reply by Benny56** [reply]
Sure.
[BuildLog.txt](https://github.com/dotnet/roslyn/files/6149926/BuildLog.txt)

**Comment by Benny56** [comment]
Meanwhile I have tried a workaround to read the file directly.
Again this works locally but not on Azure DevOps.

        public void Execute(GeneratorExecutionContext context)
        {
            try
            {
                if (context.AdditionalFiles.Length > 0)
                {
                    // Read settings file
                    SettingsXml = (from file in context.AdditionalFiles
                                   where file.Path.ToLower().EndsWith(FILENAME.ToLower())
                                   select file.GetText(context.CancellationToken)?.ToString())
                                   .FirstOrDefault();

                    // Generate multiple sources if we have a control file
                    if (SettingsXml != null) GenerateSources();
                    else context.ReportDiagnostic(Diagnostic.Create(MissingSettingsFile, Location.None, FILENAME));
                }
                else
                {
                    /********************************************************************************************** 
                     * WORKAROUND
                     * This is a workaround for Azure DevOps, when the AdditionalFiles property is not initialized.
                     * We try to find the control file by ourselves.
                     **********************************************************************************************/

                    /* Locally the full pathname is returned. On Azure DevOps however null is returned. */
                    /************************************************************************************/ 
                    var pathname = context.Compilation.Options.SourceReferenceResolver.ResolveReference(FILENAME, null);
                   
                    if (pathname != null)
                    {
                        var sourceText = context.Compilation.Options.SourceReferenceResolver.ReadText(pathname);
                        SettingsXml = sourceText.ToString();
                        GenerateSources();
                    }
                    else context.ReportDiagnostic(Diagnostic.Create(NoAdditionalFiles, Location.None, FILENAME));
                }
            }
            catch ....

        }

**Comment by Benny56** [comment, marked as answer]
It turned out that other than locally that the source generator is invoked twice. The first time when building the Android project and the second time when building the shared project. The additional files are available the second time only. Thus, the following code solves the problem:

        public void Execute(GeneratorExecutionContext context)
        {
            if (context.AdditionalFiles.Length > 0)
            {
                // Your stuff
            }
        }
